Fix Laravel 11 compability

syntafin opened the pull request:
Laravel 11 adds Carbon ^3.0, with this PR it works

syntafin committed:
Add Carbon 3.0 as dep

abdumu commented:
have you tried the last release?

syntafin commented:
Yes, doesnt work as new L11 projects comes with Carbon 3.

abdumu commented:
tests failed with this PR, work on that and i'm ready to merge it

syntafin commented:
Yes, because of a deprecation notice (float -> int):
> DEPRECATED  Implicit conversion from float -2161878.318833 to int loses precision in vendor/awssat/laravel-visits/src/DataEngines/EloquentEngine.php on line 193.

still looking into this atm.

abdumu commented:
yes, since it's late now I'll be waiting for your fix, or i will look into it tomorrow by myself.

syntafin commented:
It's late here too :D will look into it tomorrow too!

syntafin committed:
Change Type to Float

syntafin commented:
I found the problem, its caused only if you use Carbon 3, as it changed the behaivour of ``diffIn`` to strict typing:

![image](https://github.com/awssat/laravel-visits/assets/695535/d8c61afc-0949-45e3-bbbc-b5100c4ebc0b)

Changing the function ``newExpiration`` to how described in upgrade docs, only one (two) test fail: ``day_test``.

```php
protected function newExpiration($period)
    {
        try {
            $periodCarbon = $this->xHoursPeriod($period) ?? Carbon::now()->{'endOf' . Str::studly($period)}();
        } catch (Exception $e) {
            throw new Exception("Wrong period: `{$period}`! please update config/visits.php file.");
        }

        return (int) abs($periodCarbon->diffInSeconds()) + 1;
    }
```

Still ongoing to figure out that last problem.

> Time: 00:11.637, Memory: 38.50 MB
> 
> There were 2 failures:
> 
> 1) Awssat\Visits\Tests\Feature\EloquentPeriodsTest::day_test
> Failed asserting that -1.000011574074074 matches expected 1.
> 
> /home/syntafin/projects/laravel-visits/tests/Feature/PeriodsTestCase.php:95
> 
> 2) Awssat\Visits\Tests\Feature\RedisPeriodsTest::day_test
> Failed asserting that -1.0034837962962964 matches expected 1.
> 
> /home/syntafin/projects/laravel-visits/tests/Feature/PeriodsTestCase.php:95

syntafin committed:
Refactor variable names and data types across classes

Changed the variable name from 'key' to 'search' in valueList method across the engine classes to be more descriptive about its functionality. Moreover, adjusted the data types for variables in setExpiration methods from float to integer to ensure consistency throughout the code. An enhancement in Periods class has been made to make sure the difference in seconds is always an absolute integer.

syntafin commented:
Tests now should all pass, I changed them and the return type of ``newExpiration`` to use absolute integers.
May you can check it on your side and see if I didnt oversee anything :)

abdumu commented:
i pushed the last change before seeing your last comment, i would've pulled it instead of changing this myself. absolutely right. thanks for debugging this.
